### DISCUSSION TITLE
Maven Deamon (Maven 4) compatibility

### DISCUSSION THREAD

**Discussion by lppedd** [eclipse-tycho/tycho#2041] [Q&A] [answered]
Out of curiosity, I've tried running our build using Maven Deamon, which uses Maven 4.

The show stopper currently is [DefaultGraphBuilder](https://github.com/apache/maven/blob/maven-4.0.0-alpha-4/maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java), which has changed the injected constructor dependencies.  
Thus the context fails to instantiate `TychoGraphBuilder`.

Not sure if it's possible to adapt `TychoGraphBuilder` to be compatible with Maven 4 and Maven 3.x at the same time.  
Any thoughts on this? Not urgent or anything.

**Comment by laeubi** [comment, marked as answer]
Currently there is no plan to support Maven 4.

**Reply by lppedd** [reply]
Hopefully it's all already compatible, maybe apart from some dependency injection adjustments.

**Reply by laeubi** [reply]
> Hopefully it's all already compatible, maybe apart from some dependency injection adjustments.

Most likely not ... Maven 4 is only compatible for trivial plugins, and there is literally no compatibility regarding extensions (like Tycho) as you already pointed out.

**Reply by lppedd** [reply]
> no compatibility regarding extensions (like Tycho)

So basically Tycho is forced to ship a version that only targets Maven 4+. You'd have to maintain two different streams with feature parity most likely.

**Reply by laeubi** [reply]
> > no compatibility regarding extensions (like Tycho)
> 
> So basically Tycho is forced to ship a version that only targets Maven 4+. You'd have to maintain two different streams with feature parity most likely.

Each Tycho version has a required minimum maven version, and most likely no one will maintain any parallel work, so Tycho 4 will most likely target Maven 3.9.x if there are no issues with the maven-legacy used in tests (as far as I know there is still no new release for maven-testing-plugin).
